mip/options.go

```diff
@@ -1,10 +1,203 @@
 package mip
 
-import "time"
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+)
 
-// Limits on the MIP solver, such as how long it can run for.
-type Limits struct {
+// SolveOptions are options that can be cionfigured for any solver.
+type SolveOptions struct {
 	// Duration is the maximum duration of the solver. A duration limit of 0 is
 	// treated as infinity.
-	Duration time.Duration `json:"duration" usage:"maximum duration of the solver. A duration limit of 0 is treated as infinity" default:"30s"`
+	Duration time.Duration `json:"duration" usage:"Maximum duration of the solver." default:"30s"`
+	// Verbosity of the solver in the console.
+	Verbosity Verbosity `json:"verbosity" usage:"{off, low, medium, high} Verbosity of the solver in the console." default:"off"`
+	// Mip-specific options.
+	Mip MipOptions `json:"mip" usage:"Options specific to MIP problems. Linear problems do not use these options."`
+	// Control options for the specific solver.
+	Control ControlOptions `json:"control" usage:"Options to control a specific solver, as defined by the provider."`
+}
+
+// MipOptions are options specific to MIP problems. LP problems do not use
+// these options.
+type MipOptions struct {
+	// Gap stopping criteria.
+	Gap GapOptions `json:"gap" usage:"Gap stopping criteria."`
+}
+
+// GapOptions specifies the gap stopping criteria.
+type GapOptions struct {
+	// Absolute gap.
+	Absolute float64 `json:"absolute" usage:"Absolute gap stopping value. If the problem is an integer problem the solver will stop if the gap between the relaxed problem and the best found integer problem is less than this value." default:"0.0"`
+	// Relative gap.
+	Relative float64 `json:"relative" usage:"Relative gap stopping value. If the problem is an integer problem the solver will stop if the relative gap between the relaxed problem and the best found integer problem is less than this value." default:"0.0"`
+}
+
+// Verbosity specifies the level of verbosity of the back-end solver.
+type Verbosity string
+
+const (
+	// Off logs nothing.
+	Off Verbosity = "off"
+	// Low logs essentials, depends on the back-end solver.
+	Low Verbosity = "low"
+	// Medium logs essentials plus high level events,
+	// depends on the back-end solver.
+	Medium Verbosity = "medium"
+	// High logs everything the underlying logs,
+	// depends on the back-end solver.
+	High Verbosity = "high"
+)
+
+// ControlOptions allow the user to define solver-specific parameters. The
+// parameters' names and types must be known beforehand so that the correct
+// type is used.
+type ControlOptions struct {
```

Comment:
Not sure I completely see how this is working. For CPLEX and other solver the keys of the control options are integers and not strings. Is this implementation taking this into account we might have to support int to {float,int,bool,string} options?

Reply:
CPLEX works like Xpress.
- In Xpress take the [XPRSsetdblcontrol](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/optimizer/HTML/XPRSsetdblcontrol.html) function for example. The control parameter is an int and the value of course is a double.
- In CPLEX take the [CPXsetdblparam](https://www.ibm.com/docs/en/icos/22.1.1?topic=s-cpxxsetdblparam-cpxsetdblparam) function, in which the the parameter is also an int and the value is a double.

How I am handling this is I have a map that maps a string to the correct int option. This is part of the Platform PR. For any solver, the idea is to set the param name as a string and the value is set under the correct category.

Notice in the description that this works: `-solve.control.float "XPRS_BIGM=10000"`. Under the hood, `XPRS_BIGM` is translated to an int and the `XPRSsetdblcontrol` function is used as `XPRSsetdblcontrol(problem, 7018, 10000)`.